flow360/component/utils.py

```diff
@@ -145,3 +147,111 @@ def zstd_compress(file_path, output_file_path=None, compression_level=3):
     except (zstd.ZstdError, FileNotFoundError, IOError) as error:
         log.error(f"Error occurred while compressing the file: {error}")
         return None
+
+
+UNIT_VECTOR_NORM_TOLERANCE = 1e-6
+
+
+def check_unit_vector(vector: list):
+    """
+    Return whether a vector is unit vector and its magnitude
+    """
+    magnitude = 0
+    for comp in vector:
+        magnitude += comp * comp
+    magnitude = numpy.sqrt(magnitude)
+    if numpy.abs(magnitude - 1.0) < UNIT_VECTOR_NORM_TOLERANCE:
+        return (True, magnitude)
+    return (False, magnitude)
+
+
+def normalize_vector(vector, name: str):
```

Comment:
I think the vector normalisation could be pydantic field validator:

```python
    # pylint: disable=no-self-argument, protected-access
    @pd.validator("axis")
    def normalise_axis(cls, value):
      is_unit_vector, vector_norm = check_unit_vector(value)
       normalized_vector = [0, 0, 0]
       if not is_unit_vector:
           if vector_norm == 0:
               raise ValueError(f"Zero vector found")
           for dim in range(0, 3):
               normalized_vector[dim] = vector[dim] / vector_norm
           return tuple(normalized_vector)
       return vector
```
then there is no need to pass "name" as parameter and no need to call it in to_solver() method. The advantage is that error will be raised on assignment instead of on submit when we call to_solver()

see here for more details: 
https://docs.pydantic.dev/1.10/usage/validators/#pre-and-per-item-validators
https://docs.pydantic.dev/1.10/usage/validators/#reuse-validators

Reply:
Great instruction. I changed according to the example